classes/submission/SubmissionFileDAODelegate.inc.php

```diff
@@ -68,14 +68,18 @@ function insertObject($submissionFile, $sourceFile, $isUpload = false) {
 			array_unshift($params, (int) $fileId);
 		}
 
-		$this->update(
-			sprintf('INSERT INTO submission_files
-				(' . ($fileId ? 'file_id, ' : '') . 'revision, submission_id, source_file_id, source_revision, file_type, file_size, original_file_name, file_stage, date_uploaded, date_modified, viewable, uploader_user_id, assoc_type, assoc_id, genre_id, direct_sales_price, sales_type)
-				VALUES
-				(' . ($fileId ? '?, ' : '') . '?, ?, ?, ?, ?, ?, ?, ?, %s, %s, ?, ?, ?, ?, ?, ?, ?)',
-				$this->datetimeToDB($submissionFile->getDateUploaded()), $this->datetimeToDB($submissionFile->getDateModified())),
-			$params
-		);
+		$isSqlServer = Config::getVar('database', 'ms_sql');
+		$sql = sprintf(($fileId && $isSqlServer ? 'SET IDENTITY_INSERT submission_files ON;' : '') .
```

Comment:
Could you point me to something describing this aspect of SQL Server? I'm not familiar.

Reply:
The IDs for the table are generated sequentially so SQL server doesn't accept the modifications or the insertions over on this value. That's where the IDENTITY_INSERT ON/OFF comes handy in some cases. But as you said in the comment below, if we can remove the ID from the update and do a clear update instead of an insertion when the file ID is not null, than we can make the code works for every databases.

Reply:
I removed the file ID for the update section, but I can't make the same change for the insertion as the table is composed with a two columns primary key.

We need to insert a new revision each time in the table and SQL server doesn't really like it.

To make it functionnal within all databases without the use of "IDENTITY_INSERT ON/OFF", a refactor is needed to split the table in 2 (one with the file ID and revision ID as a foreign key and another one with the revision ID + associated data).